sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/aggregator/ExponentialBucketStrategy.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.aggregator;
+
+import io.opentelemetry.sdk.metrics.internal.state.ExponentialCounterFactory;
+
+/** The configuration for how to create exponential histogram buckets. */
+interface ExponentialBucketStrategy {
+  /** Constructs fresh new buckets. */
+  DoubleExponentialHistogramBuckets newBuckets();
+
+  /** Constructs "empty" count buckets using settings from previous recorded bucket. */
```

Comment:
I am trying to think if it's a safe assumption to use the scale of the previous accumulations after resetting. Using cumulative aggregation temporality, certainly it's safe. `merge()` will downscale it anyway, so it would save some CPU time. 

However for delta temporality, could there be a use case where someone might want the highest precision possible for each export?

Reply:
On a related note, I noticed diff / merge / zero are implemented as immutable operations but could they be mutable? Intuitively, always mutating a long array seems like the allocations would be fine (preallocated just once on initialization) and avoid the dynamic dispatch of the byte-short-int-long stuff. The idea does seem to at least require the scale to be preserved on reset though

Reply:
So I've exposed it as a thing we can "toy with" and generate a strategy that does NOT do this to compare against other strategies.

However, there are two common use cases we need to optimise for:

1. High throughput traffic in a distribution
2. Low throughput traffic in a distribution

Given our instrumentation, we're likely to see both of these.  For 2 we want smallest amount of memory consumption, for 1 we want high performance.   The idea behind saving Scale (and eventually integer array size) is so that we can preserve as much performance as possible for #1 by shifting distributions from use cases #2 to #1 as we "detect" high throughput.

Reply:
> could they be mutable?
@anuraaga I asked this initially when writing those operations, and it seemed that accumulations need to be immutable the way the code is at the current moment. I created an issue to allow mutable accumulations: https://github.com/open-telemetry/opentelemetry-java/issues/3766

Reply:
The short answer on mutable accumulation is "no".  They're shared between exporters now, so mutating would (initially) lead to bugs.  I'm hoping to do some more optimisation work toward that end,specifically making sure none of it lands on the hot-path.   IIRC - all `merge`/`diff` operations are already off the 'hot' path and only performed during collection/export.

Reply:
If we're able to eventually make them mutable, I think we'd be able to make the histogram aggregations no-alloc which should allow removing `AdaptingIntegerArray` as the memory overhead would be fixed, or at least trim it down to being bimorphic which Java deals pretty well with. Will be a good area to explore as that class really just feels like a "root of all evil"

Reply:
I'm not really sure how this particular optimisation is any more evil than hand-rolling Protobuf serialization, from both a maintenance and "scare" sense.   Can you elaborate on primary concerns, especially given the benchmark showing performance?  I'm happy to add testing around this, but I definitely don't see it as anywhere near as scary as hand-rolled serialization.

Related, I'm really not sure what you mean about mutable + optimising.  Could you lay out that scenario/idea?  I'd be happy to try it out.

I'm afraid, though, that in practice the we need to optimise for the two scenarios I listed above.   "Eventually" using consistent amount of memory may not be good enough if the overhead per-timeseries is still too high.   However, I'm not sure I understand what kind of mutation you're proposing, and I'm not sure if I've accurately described *why* using the adapting array plays to the current use case we have for histograms so well.

Reply:
The protobuf optimizations are pretty deterministic - we can reason out the difference in the old and new code to know where performance will come from. The adapting arrays are based on heuristics - in particular, we would expect direct array access to always perform quite a bit better than an access through a megamorphic switch statement, so there is a clear tradeoff involved. This makes it much scarier and microbenchmarks will never be able to give the full picture when relying on heuristics and assumptions. Gradually going up in complexity tends to at least be a bit less scary, for example MapCounter -> `long[]` circular buffer -> then `AdaptingIntegerArray` cirulcar buffer.

> Related, I'm really not sure what you mean about mutable + optimizing. Could you lay out that scenario/idea? I'd be happy to try it out.

As far as I could tell, if all the operations on the bucket counts were mutating an array, including zeroing at the end of collection, there wouldn't be a need to reallocate the bucket count storage during operations / collection. If so, it would seem to incentivize only having one type, `long[]` or at most two types since Java could still optimize bimorphic code paths well. If it's not possible to remove the allocations, though, then it is what it is and raises the importance of the adapting integers

Reply:
> Gradually going up in complexity tends to at least be a bit less scary, for example MapCounter -> long[] circular buffer -> then AdaptingIntegerArray cirulcar buffer.

I shared (a few weeks back, when I first sent the perf benchmarks) a branch I had where I had all three of these.

The `long[]` counter is the clear winner for raw CPU performance.  There is a minor performance degredation going to the `Adapting` array (as you call out, due to the switch).   However, as shown in this PR, the Adapting use case saves a good bit of RAM in addition to remaining faster than `MapCounter`.   Given the differential, and the sheer # of timeseries I see coming out of instrumentation histograms today, I think this RAM optimisation (or something like it) is worth it.   Perhaps there's a different mechanism we can use to change the array implementation that would be more ideal in leading to (eventual) monomorphic optimisation?



> As far as I could tell, if all the operations on the bucket counts were mutating an array, including zeroing at the end of collection, there wouldn't be a need to reallocate the bucket count storage during operations / collection. If so, it would seem to incentivize only having one type, long[] or at most two types since Java could still optimize bimorphic code paths well. If it's not possible to remove the allocations, though, then it is what it is and raises the importance of the adapting integers

We could actually do this today.   These counters *are* mutable, we only need them immutable in the collection route.   Instead of "new"-ing up the counters, we can just create copies to expose externally and then call `.zero`.   Happy to do that in this PR or another.

Reply:
> I think this RAM optimisation (or something like it) is worth it.

300 buckets * 8 is 2.4K per histogram, 2000 histograms in 5MB seems like a lot of histograms for what in Java is not a lot of RAM. 

5MB allocations per cycle on the other hand becomes more cause for concern which is where the point of mutability comes in. If reallocations an be avoided, then I really feel the case for an adapting array becomes weaker.

Anyways I already approved since don't care about this code that much. It's just I've generally seen situations that jump at the most complex solution to not scale back to a simpler one even if that happened to be more appropriate, going in steps tends to work better to not get too deep into a complexity hole.